File: apps/api/app/schemas/report.py
"""Report schemas - Pydantic models for API contracts."""

from datetime import date, datetime
from decimal import Decimal
from enum import Enum
from typing import Any, Optional
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from app.schemas.base import BaseSchema


# Enums
class ReportType(str, Enum):
    """Type of report to generate."""

    # Financial reports
    DRE = "dre"
    FLUXO_CAIXA = "fluxo_caixa"
    LIVRO_CAIXA = "livro_caixa"
    RECEITAS_CLIENTE = "receitas_cliente"
    DESPESAS_CATEGORIA = "despesas_categoria"
    PROJECAO_FLUXO = "projecao_fluxo"
    KPIS = "kpis"

    # Operational reports
    CLIENTES = "clientes"
    OBRIGACOES = "obrigacoes"
    LICENCAS = "licencas"
    AUDITORIA = "auditoria"


class ReportFormat(str, Enum):
    """Export format for reports."""

    PDF = "pdf"
    CSV = "csv"
    XLS = "xls"


class ReportStatus(str, Enum):
    """Status of report generation."""

    PENDING = "pending"
    COMPLETED = "completed"
    FAILED = "failed"


class ChartType(str, Enum):
    """Type of chart for visualization."""

    LINE = "line"
    BAR = "bar"
    PIE = "pie"
    DONUT = "donut"
    AREA = "area"
    TABLE = "table"


# Request schemas
class ReportFilterRequest(BaseSchema):
    """Base schema for report filters."""

    period_start: date = Field(..., description="Start date of the period")
    period_end: date = Field(..., description="End date of the period")
    client_ids: Optional[list[UUID]] = Field(None, description="Filter by specific clients")
    report_type: ReportType = Field(..., description="Type of report to generate")

    @field_validator("period_end")
    def validate_period(cls, v: date, info) -> date:
        """Ensure period_end is after period_start."""
        if "period_start" in info.data and v < info.data["period_start"]:
            raise ValueError("period_end must be after period_start")
        return v

    class Config:
        json_schema_extra = {
            "example": {
                "period_start": "2025-01-01",
                "period_end": "2025-12-31",
                "client_ids": None,
                "report_type": "dre",
            }
        }


class ReportCustomization(BaseSchema):
    """Schema for report customization options."""

    fields_to_include: Optional[list[str]] = Field(
        None, description="Specific fields to include in the report"
    )
    group_by: Optional[str] = Field(None, description="Field to group data by")
    sort_by: Optional[str] = Field(None, description="Field to sort by")
    sort_direction: Optional[str] = Field("asc", description="Sort direction (asc/desc)")
    chart_types: Optional[list[ChartType]] = Field(
        None, description="Types of charts to include"
    )
    include_summary: bool = Field(True, description="Include summary section")
    include_charts: bool = Field(True, description="Include charts/visualizations")

    class Config:
        json_schema_extra = {
            "example": {
                "fields_to_include": ["data", "valor", "cliente"],
                "group_by": "cliente",
                "sort_by": "valor",
                "sort_direction": "desc",
                "chart_types": ["bar", "pie"],
                "include_summary": True,
                "include_charts": True,
            }
        }


class ReportPreviewRequest(BaseSchema):
    """Request schema for generating report preview."""

    report_type: ReportType = Field(..., description="Type of report")
    filters: ReportFilterRequest = Field(..., description="Report filters")
    customizations: Optional[ReportCustomization] = Field(
        None, description="Customization options"
    )

    class Config:
        json_schema_extra = {
            "example": {
                "report_type": "dre",
                "filters": {
                    "period_start": "2025-01-01",
                    "period_end": "2025-12-31",
                    "client_ids": None,
                    "report_type": "dre",
                },
                "customizations": {
                    "include_summary": True,
                    "include_charts": True,
                },
            }
        }


class ReportExportRequest(BaseSchema):
    """Request schema for exporting report."""

    report_type: ReportType = Field(..., description="Type of report")
    format: ReportFormat = Field(..., description="Export format")
    filters: ReportFilterRequest = Field(..., description="Report filters")
    customizations: Optional[ReportCustomization] = Field(
        None, description="Customization options"
    )
    filename: Optional[str] = Field(None, max_length=200, description="Custom filename")
    save_as_template: bool = Field(False, description="Save as reusable template")

    class Config:
        json_schema_extra = {
            "example": {
                "report_type": "dre",
                "format": "pdf",
                "filters": {
                    "period_start": "2025-01-01",
                    "period_end": "2025-12-31",
                    "client_ids": None,
                    "report_type": "dre",
                },
                "customizations": None,
                "filename": "DRE_2025",
                "save_as_template": False,
            }
        }


# Response schemas
class ReportPreviewResponse(BaseSchema):
    """Response schema for report preview."""

    report_type: ReportType
    data: dict[str, Any] = Field(..., description="Report data")
    charts_config: Optional[list[dict[str, Any]]] = Field(
        None, description="Chart configuration"
    )
    summary: Optional[dict[str, Any]] = Field(None, description="Report summary")
    generated_at: datetime = Field(default_factory=datetime.utcnow)
    record_count: int = Field(..., description="Number of records in the report")


class ReportExportResponse(BaseSchema):
    """Response schema for report export."""

    report_id: UUID = Field(..., description="Report history ID")
    file_url: str = Field(..., description="URL to download the file")
    file_name: str = Field(..., description="Generated filename")
    file_size: int = Field(..., description="File size in bytes")
    format: ReportFormat
    generated_at: datetime
    expires_at: datetime = Field(..., description="File expiration datetime")


class ChartConfig(BaseSchema):
    """Configuration for a chart."""

    type: ChartType
    title: str
    data_key: str
    x_axis_key: Optional[str] = None
    y_axis_key: Optional[str] = None
    color_scheme: Optional[list[str]] = None


# Template schemas
class ReportTemplateCreate(BaseSchema):
    """Schema for creating a report template."""

    name: str = Field(..., min_length=1, max_length=200, description="Template name")
    description: Optional[str] = Field(None, max_length=500, description="Description")
    report_type: ReportType = Field(..., description="Type of report")
    default_filters: dict[str, Any] = Field(..., description="Default filters")
    default_customizations: Optional[dict[str, Any]] = Field(
        None, description="Default customizations"
    )

    class Config:
        json_schema_extra = {
            "example": {
                "name": "DRE Mensal",
                "description": "Demonstrativo de Resultados mensal padrão",
                "report_type": "dre",
                "default_filters": {
                    "period_start": "2025-01-01",
                    "period_end": "2025-12-31",
                },
                "default_customizations": {
                    "include_summary": True,
                    "include_charts": True,
                },
            }
        }


class ReportTemplateUpdate(BaseSchema):
    """Schema for updating a report template."""

    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=500)
    default_filters: Optional[dict[str, Any]] = None
    default_customizations: Optional[dict[str, Any]] = None


class ReportTemplateResponse(BaseSchema):
    """Schema for report template response."""

    id: UUID
    name: str
    description: Optional[str]
    report_type: ReportType
    default_filters: dict[str, Any]
    default_customizations: Optional[dict[str, Any]]
    is_system: bool
    created_by_id: UUID
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# History schemas
class ReportHistoryResponse(BaseSchema):
    """Schema for report history response."""

    id: UUID
    template_id: Optional[UUID]
    user_id: UUID
    report_type: ReportType
    filters_used: dict[str, Any]
    format: ReportFormat
    file_path: Optional[str]
    file_size: Optional[int]
    generated_at: datetime
    expires_at: Optional[datetime]
    status: ReportStatus

    class Config:
        from_attributes = True


class ReportHistoryListResponse(BaseSchema):
    """Paginated report history response."""

    items: list[ReportHistoryResponse]
    total: int
    page: int
    size: int
    pages: int


# Report type info
class ReportTypeInfo(BaseSchema):
    """Information about a report type."""

    type: ReportType
    name: str
    description: str
    category: str  # "financeiro" or "operacional"
    supports_customization: bool
    supported_charts: list[ChartType]
    required_permissions: Optional[list[str]] = None


class ReportTypesListResponse(BaseSchema):
    """List of available report types."""

    types: list[ReportTypeInfo]


# Financial report specific schemas
class DREReportData(BaseSchema):
    """DRE Report data structure."""

    class DRERow(BaseSchema):
        categoria: str
        valor: Decimal
        percentual: Optional[float] = None

    receitas: list[DRERow] = Field(..., description="Revenue items")
    despesas: list[DRERow] = Field(..., description="Expense items")
    receita_total: Decimal
    despesa_total: Decimal
    resultado_liquido: Decimal
    margem_lucro: float


class CashFlowReportData(BaseSchema):
    """Cash Flow Report data structure."""

    class CashFlowPeriod(BaseSchema):
        periodo: str  # YYYY-MM
        entradas: Decimal
        saidas: Decimal
        saldo_inicial: Decimal
        saldo_final: Decimal

    periods: list[CashFlowPeriod]
    total_entradas: Decimal
    total_saidas: Decimal
    saldo_final_periodo: Decimal


class CashBookReportData(BaseSchema):
    """Cash Book Report data structure."""

    class CashBookEntry(BaseSchema):
        data: date
        tipo: str  # "entrada" or "saida"
        descricao: str
        valor: Decimal
        saldo_acumulado: Decimal

    entries: list[CashBookEntry]
    saldo_inicial: Decimal
    saldo_final: Decimal
    total_entradas: Decimal
    total_saidas: Decimal


class RevenueByClientReportData(BaseSchema):
    """Revenue by Client Report data structure."""

    class ClientRevenue(BaseSchema):
        client_id: UUID
        client_name: str
        client_cnpj: Optional[str]
        total_receita: Decimal
        percentual_total: float

    clients: list[ClientRevenue]
    total_receita: Decimal


class ExpensesByCategoryReportData(BaseSchema):
    """Expenses by Category Report data structure."""

    class CategoryExpense(BaseSchema):
        categoria: str
        total: Decimal
        percentual_total: float

    categories: list[CategoryExpense]
    total_despesas: Decimal


class CashFlowProjectionReportData(BaseSchema):
    """Cash Flow Projection Report data structure."""

    class ProjectionPeriod(BaseSchema):
        periodo: str  # YYYY-MM
        cenario_otimista: Decimal
        cenario_realista: Decimal
        cenario_pessimista: Decimal

    periods: list[ProjectionPeriod]
    metodo_projecao: str
    base_historico_meses: int


class KPIReportData(BaseSchema):
    """Financial KPIs Report data structure."""

    margem_lucro: float = Field(..., description="Profit margin %")
    percentual_despesas_fixas: float = Field(..., description="Fixed expenses % of revenue")
    taxa_inadimplencia: float = Field(..., description="Default rate %")
    ticket_medio: Decimal = Field(..., description="Average ticket per client")
    crescimento_mom: float = Field(..., description="Month-over-month growth %")
    crescimento_yoy: float = Field(..., description="Year-over-year growth %")
    roi: Optional[float] = Field(None, description="Return on investment %")


# Operational report schemas
class ClientsReportData(BaseSchema):
    """Clients Report data structure."""

    class ClientInfo(BaseSchema):
        id: UUID
        razao_social: str
        cnpj: str
        email: Optional[str]
        status: str
        honorarios: Decimal
        total_pendente: Decimal
        total_atrasado: Decimal

    clients: list[ClientInfo]
    total_clientes: int
    total_honorarios: Decimal


class ObligationsReportData(BaseSchema):
    """Obligations Report data structure."""

    compliance_rate: float = Field(..., description="% of obligations completed on time")
    total_obligations: int
    pending: int
    completed: int
    overdue: int
    cancelled: int


class LicensesReportData(BaseSchema):
    """Licenses Report data structure."""

    total_licenses: int
    active: int
    expiring_soon: int  # < 30 days
    expired: int
    renewals_in_period: int


class AuditReportData(BaseSchema):
    """Audit Report data structure."""

    total_actions: int
    actions_by_user: dict[str, int]
    actions_by_module: dict[str, int]
    actions_by_type: dict[str, int]

File: apps/web/src/types/report.ts
/**
 * Report types - TypeScript interfaces for frontend
 */

// Enums
export enum ReportType {
  // Strategic reports
  GERAL = 'geral',

  // Financial reports
  DRE = 'dre',
  FLUXO_CAIXA = 'fluxo_caixa',
  LIVRO_CAIXA = 'livro_caixa',
  RECEITAS_CLIENTE = 'receitas_cliente',
  DESPESAS_CATEGORIA = 'despesas_categoria',
  PROJECAO_FLUXO = 'projecao_fluxo',
  KPIS = 'kpis',

  // Operational reports
  CLIENTES = 'clientes',
  OBRIGACOES = 'obrigacoes',
  LICENCAS = 'licencas',
  AUDITORIA = 'auditoria',
}

export enum ReportFormat {
  PDF = 'pdf',
  CSV = 'csv',
  XLS = 'xls',
}

export enum ReportStatus {
  PENDING = 'pending',
  COMPLETED = 'completed',
  FAILED = 'failed',
}

export enum ChartType {
  LINE = 'line',
  BAR = 'bar',
  PIE = 'pie',
  DONUT = 'donut',
  AREA = 'area',
  TABLE = 'table',
}

// Request interfaces
export interface ReportFilterRequest {
  period_start: string; // ISO date string
  period_end: string; // ISO date string
  client_ids?: string[] | null;
  status?: string | null;
  statuses?: string[] | null;
  report_type: ReportType;
}

export interface ReportCustomization {
  fields_to_include?: string[] | null;
  group_by?: string | null;
  sort_by?: string | null;
  sort_direction?: 'asc' | 'desc';
  chart_types?: ChartType[] | null;
  include_summary?: boolean;
  include_charts?: boolean;
}

export interface ReportPreviewRequest {
  report_type: ReportType;
  filters: ReportFilterRequest;
  customizations?: ReportCustomization | null;
}

export interface ReportExportRequest {
  report_type: ReportType;
  format: ReportFormat;
  filters: ReportFilterRequest;
  customizations?: ReportCustomization | null;
  filename?: string | null;
  save_as_template?: boolean;
}

// Response interfaces
export interface ReportPreviewResponse {
  report_type: ReportType;
  data: Record<string, any>;
  charts_config?: Array<Record<string, any>> | null;
  summary?: Record<string, any> | null;
  generated_at: string; // ISO datetime string
  record_count: number;
}

export interface ReportExportResponse {
  report_id: string;
  file_url: string;
  file_name: string;
  file_size: number;
  format: ReportFormat;
  generated_at: string; // ISO datetime string
  expires_at: string; // ISO datetime string
}

export interface ChartConfig {
  type: ChartType;
  title: string;
  data_key: string;
  x_axis_key?: string | null;
  y_axis_key?: string | null;
  color_scheme?: string[] | null;
}

// Template interfaces
export interface ReportTemplateCreate {
  name: string;
  description?: string | null;
  report_type: ReportType;
  default_filters: Record<string, any>;
  default_customizations?: Record<string, any> | null;
}

export interface ReportTemplateUpdate {
  name?: string;
  description?: string;
  default_filters?: Record<string, any>;
  default_customizations?: Record<string, any>;
}

export interface ReportTemplate {
  id: string;
  name: string;
  description?: string | null;
  report_type: ReportType;
  default_filters: Record<string, any>;
  default_customizations?: Record<string, any> | null;
  is_system: boolean;
  created_by_id: string;
  created_at: string; // ISO datetime string
  updated_at: string; // ISO datetime string
}

// History interfaces
export interface ReportHistory {
  id: string;
  template_id?: string | null;
  user_id: string;
  report_type: ReportType;
  filters_used: Record<string, any>;
  format: ReportFormat;
  file_path?: string | null;
  file_size?: number | null;
  generated_at: string; // ISO datetime string
  deleted_at?: string | null; // ISO datetime string
  expires_at?: string | null; // ISO datetime string
  status: ReportStatus;
}

export interface ReportHistoryListResponse {
  items: ReportHistory[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

// Report type info
export interface ReportTypeInfo {
  type: ReportType;
  name: string;
  description: string;
  category: 'financeiro' | 'operacional';
  supports_customization: boolean;
  supported_charts: ChartType[];
  required_permissions?: string[] | null;
}

export interface ReportTypesListResponse {
  types: ReportTypeInfo[];
}

// Financial report data structures
export interface DRERow {
  categoria: string;
  valor: number;
  percentual?: number | null;
}

export interface DREReportData {
  receitas: DRERow[];
  despesas: DRERow[];
  receita_total: number;
  despesa_total: number;
  resultado_liquido: number;
  margem_lucro: number;
}

export interface CashFlowPeriod {
  periodo: string; // YYYY-MM
  entradas: number;
  saidas: number;
  saldo_inicial: number;
  saldo_final: number;
}

export interface CashFlowReportData {
  periods: CashFlowPeriod[];
  total_entradas: number;
  total_saidas: number;
  saldo_final_periodo: number;
}

export interface CashBookEntry {
  data: string; // ISO date string
  tipo: 'entrada' | 'saida';
  descricao: string;
  valor: number;
  saldo_acumulado: number;
}

export interface CashBookReportData {
  entries: CashBookEntry[];
  saldo_inicial: number;
  saldo_final: number;
  total_entradas: number;
  total_saidas: number;
}

export interface ClientRevenue {
  client_id: string;
  client_name: string;
  client_cnpj?: string | null;
  total_receita: number;
  percentual_total: number;
}

export interface RevenueByClientReportData {
  clients: ClientRevenue[];
  total_receita: number;
}

export interface CategoryExpense {
  categoria: string;
  total: number;
  percentual_total: number;
}

export interface ExpensesByCategoryReportData {
  categories: CategoryExpense[];
  total_despesas: number;
}

export interface ProjectionPeriod {
  periodo: string; // YYYY-MM
  cenario_otimista: number;
  cenario_realista: number;
  cenario_pessimista: number;
}

export interface CashFlowProjectionReportData {
  periods: ProjectionPeriod[];
  metodo_projecao: string;
  base_historico_meses: number;
}

export interface KPIReportData {
  receita_total?: number;
  despesa_total?: number;
  resultado_liquido?: number;
  total_pendente?: number;
  total_atrasado?: number;
  total_receber?: number;
  total_clientes?: number;
  total_transacoes?: number;
  total_transacoes_atrasadas?: number;
  margem_lucro: number;
  percentual_despesas_fixas: number;
  taxa_inadimplencia: number;
  ticket_medio: number;
  crescimento_mom: number;
  crescimento_yoy: number;
  roi?: number | null;
}

export interface GeneralReportCompany {
  escopo: 'escritorio' | 'cliente' | 'consolidado';
  id?: string | null;
  nome: string;
  razao_social?: string | null;
  cnpj?: string | null;
  email?: string | null;
  telefone?: string | null;
  endereco?: string | null;
}

export interface GeneralMonthlyEvolution {
  competencia: string;
  receita_total: number;
  despesa_total: number;
  resultado_liquido: number;
  margem_lucro: number;
}

export interface GeneralProjectionPeriod {
  competencia: string;
  previsao_receita: number;
  previsao_despesa: number;
  previsao_resultado: number;
}

export interface GeneralReportData {
  empresa: GeneralReportCompany;
  resumo_financeiro: {
    receita_total: number;
    despesa_total: number;
    resultado_liquido: number;
    margem_lucro: number;
  };
  dre_simplificada: DREReportData;
  kpis: {
    margem_operacional: number;
    resultado_liquido: number;
    receita_total: number;
    despesa_total: number;
  };
  evolucao_mensal: GeneralMonthlyEvolution[];
  analises: {
    principais_receitas: DRERow[];
    principais_despesas: DRERow[];
  };
  projecoes: {
    metodo_projecao: string;
    base_historico_meses: number;
    periodos: GeneralProjectionPeriod[];
  };
}

// Operational report data structures
export interface ClientInfo {
  id: string;
  razao_social: string;
  nome_fantasia?: string | null;
  cnpj: string;
  email?: string | null;
  status: string;
  regime_tributario: string;
  honorarios: number;
  total_pendente: number;
  total_atrasado: number;
}

export interface ClientsReportData {
  clients: ClientInfo[];
  total_clientes: number;
  total_clientes_ativos: number;
  total_honorarios: number;
  por_regime: Array<{ regime: string; total: number }>;
  por_status: Array<{ status: string; total: number }>;
}

export interface ObligationReportItem {
  id: string;
  client_id: string;
  client_name: string;
  client_razao_social: string;
  client_cnpj: string;
  obligation_name: string;
  obligation_code: string;
  status: string;
  competencia: string;
  due_date: string;
  priority: string;
  completed_at?: string | null;
}

export interface ObligationsReportData {
  compliance_rate: number;
  total_obligations: number;
  pending: number;
  completed: number;
  overdue: number;
  cancelled: number;
  totais_por_status: Array<{ status: string; total: number }>;
  obligations: ObligationReportItem[];
}

export interface LicensesReportData {
  total_licenses: number;
  active: number;
  expiring_soon: number;
  expired: number;
  renewals_in_period: number;
}

export interface AuditReportData {
  total_actions: number;
  actions_by_user: Record<string, number>;
  actions_by_module: Record<string, number>;
  actions_by_type: Record<string, number>;
}

// Helper functions
export function getReportTypeLabel(type: ReportType): string {
  const labels: Record<ReportType, string> = {
    [ReportType.GERAL]: 'Relatório Geral',
    [ReportType.DRE]: 'Demonstrativo de Resultados (DRE)',
    [ReportType.FLUXO_CAIXA]: 'Fluxo de Caixa',
    [ReportType.LIVRO_CAIXA]: 'Livro Caixa',
    [ReportType.RECEITAS_CLIENTE]: 'Receitas por Cliente',
    [ReportType.DESPESAS_CATEGORIA]: 'Despesas por Categoria',
    [ReportType.PROJECAO_FLUXO]: 'Projeção de Fluxo de Caixa',
    [ReportType.KPIS]: 'Indicadores Financeiros (KPIs)',
    [ReportType.CLIENTES]: 'Relatório de Clientes',
    [ReportType.OBRIGACOES]: 'Relatório de Obrigações',
    [ReportType.LICENCAS]: 'Relatório de Licenças',
    [ReportType.AUDITORIA]: 'Relatório de Auditoria',
  };
  return labels[type];
}

export function getReportTypeCategory(type: ReportType): 'financeiro' | 'operacional' {
  const financialTypes = [
    ReportType.GERAL,
    ReportType.DRE,
    ReportType.FLUXO_CAIXA,
    ReportType.LIVRO_CAIXA,
    ReportType.RECEITAS_CLIENTE,
    ReportType.DESPESAS_CATEGORIA,
    ReportType.PROJECAO_FLUXO,
    ReportType.KPIS,
  ];
  return financialTypes.includes(type) ? 'financeiro' : 'operacional';
}

export function getReportFormatLabel(format: ReportFormat): string {
  const labels: Record<ReportFormat, string> = {
    [ReportFormat.PDF]: 'PDF',
    [ReportFormat.CSV]: 'CSV',
    [ReportFormat.XLS]: 'XLS',
  };
  return labels[format];
}

export function getReportStatusLabel(status: ReportStatus): string {
  const labels: Record<ReportStatus, string> = {
    [ReportStatus.PENDING]: 'Gerando',
    [ReportStatus.COMPLETED]: 'Concluído',
    [ReportStatus.FAILED]: 'Falhou',
  };
  return labels[status];
}

export function getReportStatusColor(
  status: ReportStatus
): 'default' | 'primary' | 'success' | 'warning' | 'danger' {
  const colors: Record<ReportStatus, 'default' | 'primary' | 'success' | 'warning' | 'danger'> = {
    [ReportStatus.PENDING]: 'primary',
    [ReportStatus.COMPLETED]: 'success',
    [ReportStatus.FAILED]: 'danger',
  };
  return colors[status];
}
